src/data_storage/import/import_storage.ts

```diff
@@ -181,14 +185,27 @@ export default class ImportStorage extends PostgresStorage {
                             LEFT JOIN data_staging ON data_staging.import_id = imports.id
                    WHERE imports.data_source_id = $1
                    GROUP BY imports.id`,
-            values: [importAdapterID]
+            values: [dataSourceID]
         }
     }
 
-    private static listReady(importAdapterID: string, offset:number, limit:number): QueryConfig {
+    private static listReadyStatement(dataSourceID: string, offset:number, limit:number): QueryConfig {
         return {
             text: `SELECT * FROM imports WHERE data_source_id = $1 AND status = 'ready' OFFSET $2 LIMIT $3`,
-            values: [importAdapterID, offset, limit]
+            values: [dataSourceID, offset, limit]
+        }
+    }
+
+    private static listIncompleteWithUninsertedDataStatement(dataSourceID: string): QueryConfig {
+        return {
+            text: `SELECT imports.*
+            FROM imports
+            WHERE imports.status <> 'completed'
+            AND imports.data_source_id = $1
+            AND EXISTS (SELECT * FROM data_staging WHERE data_staging.import_id = imports.id AND data_staging.inserted_at IS NULL)
```

Comment:
Can you explain to me why both of these AND EXISTS statements are necessary?

Reply:
The AND EXISTS statements insure that we're only getting imports back which have uninserted data records. The other AND EXISTS on the type mapping insures that we only return type mappings which are active and have type mapping transformation records.